src/boot/efi/boot.c

```diff
@@ -2268,6 +2335,115 @@ static VOID config_write_entries_to_variable(Config *config) {
+                        return err;
+                }
+
+                err = efivar_set_raw(&sb_vars[i].vendor, sb_vars[i].name, buffer, size, sb_vars_opts);
+                if (EFI_ERROR(err)) {
+                        Print(L"Failed to write %s secure boot variable: %r\n", sb_vars[i].name, err);
+                        return err;
+                }
+        }
+
+        // if we loaded the signing keys successfully then we remove all the
+        // signing keys entries from  the menu as the system is now locked down.
+        config_remove_by_id_startswith(config, L"signing-keys-");
+
+        Print(L"Finished loading signature keys from %s.\n", keys_dir);
+        uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+        return err;
+}
+
+static EFI_STATUS secure_boot_discover_keys(Config *config, EFI_FILE *root_dir) {
+        EFI_STATUS err = EFI_SUCCESS;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE keys_basedir = NULL;
+
+        if (!setup_mode_enabled())
+                return EFI_SUCCESS;
+        
+        Print(L"Setup mode is enabled.\n");
+
+        // if auto enrollement is activated (this is the default), we try to
+        // load keys straight from the /loader/keys/auto location. If it fails 
+        // (e.g. there is no such directory), we still proceed to add the
+        // entries to the menu (including the auto one)
+        if (config->auto_enroll == true) {
+                Print(L"Auto enrollement is enabled.\n");
+                err = secure_boot_enroll_from(config, root_dir, L"\\loader\\keys\\auto");
+
+                if (!EFI_ERROR(err))
+                        return EFI_SUCCESS;
+                
+                Print(L"Failed automatic enrollement.\n");
+                uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+        }
+
+        err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &keys_basedir, (CHAR16*) L"\\loader\\keys", EFI_FILE_MODE_READ, 0ULL);
+        if (EFI_ERROR(err)) {
+                Print(L"Failed opening directory /loader/keys.\n");
+                uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+                return err;
+        }
+
+        for (;;) {
+                _cleanup_freepool_ CHAR16 *keys_dir = NULL;
+                CHAR16 buf[256];
+                UINTN bufsize;
+                EFI_FILE_INFO *f;
+                bufsize = sizeof(buf);
+
+                err = uefi_call_wrapper(root_dir->Read, 3, keys_basedir, &bufsize, buf);
+                if (bufsize == 0 || EFI_ERROR(err))
+                        break;
+
+                f = (EFI_FILE_INFO *) buf;
+                if (f->FileName[0] == '.')
+                        continue;
+
+                if (!(f->Attribute & EFI_FILE_DIRECTORY))
+                        continue;
+
+                keys_dir = PoolPrint(L"\\loader\\keys\\%s", f->FileName);
+                Print(L"Found directory %s, generating boot entry.\n", keys_dir);
```

Comment:
And yet another message not needed. We can tell this was done by the fact we have another entry in the menu.